url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/graph/suffixtree/SuffixTree.java

```diff
@@ -0,0 +1,1018 @@
+	      // print the node itself 
+	      while ( start <= end )
+	      {
+	         System.out.print( (char)treeString[start] );
+	         start++;
+	      }
+	      if ( debug )
+	         System.out.print("  \t\t\t("+node1.edgeLabelStart+","
+	        	+end+" | "+node1.pathPosition+")");
+	      System.out.print("\n");
+	   }
+	   // recursive call for all node1's sons
+	   while ( node2 != null )
+	   {
+	      printNode( node2, depth+1 );
+	      node2 = node2.rightSibling;
+	   }
+	}
+	/**
+	 * This function prints the full path of a node, starting from the root. It 
+	 * calls itself recursively and than prints the last edge.
+	 * @param node the node to be printed
+	 */
+	void printFullNode( Node node )
+	{
+	   int start, end;
+	   if ( node != null )
+	   {
+		   // calculating the beginning and ending of the last edge
+		   start = node.edgeLabelStart;
+		   end = getNodeLabelEnd( node );
+		   
+		   // stopping condition - the root
+		   if ( node.father != root )
+		      printFullNode( node.father );
+		   // print the last edge
+		   while ( start <= end )
+		   {
+		      System.out.print(treeString[start]);
+		      start++;
+		   }
+	   }
+	}
+	/**
+	 * Self test of the tree - search for all substrings of the main string. See 
+	 * testing paragraph in the readme.txt file.
+	 * @return true for success false for failure
+	 */
+	boolean selfTest()
+	{
+	   int k,j,i;
+	   // loop for all the prefixes of the tree source string
+	   for ( k=1; k<length; k++ )
+	   {
+	      // loop for each suffix of each prefix
+	      for ( j=1; j<=k; j++ )
+	      {
+	         // search for the current suffix in the tree 
+	    	 int len = k-j+1;
+	    	 char[] test = new char[len];
+	    	 for ( int m=0;m<len;m++ )
+	    		 test[m] = treeString[j+m];
+	         i = findSubstring( test );
+	         if ( i == stError )
+	         {
+	            System.out.println("\n\nTest Results: Fail in string ("+j+","+k+").\n");
+	            return false;
+	         }
+	      }
+	   }
+	   // if we are here no search has failed and the test passed successfully
+	   System.out.println("\n\nTest Results: Success.\n");
+	   return true;
+	}
+    static Entry entries = null;
+    static void appendEntry( Entry e )
+    {
+        if ( entries == null )
+            entries = e;
+        else
+        {
+            Entry current = entries;
+            while ( current.next != null )
+            {
+                current = current.next;
+            }
+            current.next = e;
+        }
+    }
+	/** 
+	 * Test routine
+	 * @param args - ignored
+	 * @throws MVDException
+	 */
+	public static void main( String[] args ) throws MVDException
+	{
+		if ( args.length == 1 )
+        {
+            File file = new File( args[0] );
+            File[] files = file.listFiles();
+            try
+            {
+                for ( int i=0;i<files.length;i++ )
+                {
+                    FileInputStream fis = new FileInputStream(files[i]);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fis*. The resource is referenced at line 953. The resource is closed at line 954. There are other execution paths that do not contain closure statements, for example, when FileInputStream.read throws an exception. Ensure *fis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).